src/persist-client/src/internal/compact.rs

```diff
@@ -736,6 +736,7 @@ where
                 key_vec.extend_from_slice(k);
                 val_vec.clear();
                 val_vec.extend_from_slice(v);
+                crate::batch::validate_schema(&real_schemas, &key_vec, &val_vec, None, None);
```

Comment:
🤔 Interesting case! Bailing in CI seems like a good thing here, but I'm not sure what the desired "production" behaviour would be. What if the user changes the schema (either the data or the trait impl) such that keys / values that used to pass validation now fail... are they allowed to do that? If not, can we prevent it? If so, what do we do with that now-bad data?

Reply:
Ah, I think this is an artifact of me writing this before we had our talk the other day. I've been thinking that compaction would need to be migrating all the of input data to some consistent output schema before being added to the batch builder, even if it came in with mixed schemas. I know now that you've potentially been picturing something different.

In the short term, this check more than the other is what detected the two existing `RelationDesc::empty` usages, so IMO it's worth merging. Happy for us all to continue to discuss async how compaction plays with schema migrations and then adjust this (or rip it out) when we make a decision there and do the schema migration work.

Reply:
FWIW I think this particular musing applies to any version that relies on calling user logic during compaction, including some of the stuff I was tossing around. It seems pretty clear to me how to signal an error to the caller when a user batch doesn't validate correctly - we have a natural path to percolate the error up that isn't just taking down the replica. But validation errors that we discover during compaction don't have anywhere obvious to go...

> [...] so IMO it's worth merging.

Yeah, and since my original top-level comment / approval may have been insufficiently clear - I think this is cool to merge! We may evolve it along as our expectations around the schema stuff becomes more clear, but that seems normal...